filetracker/scripts/recover.py

```diff
@@ -0,0 +1,130 @@
+    parser.add_argument('root', help='root directory of filetracker storage')
+    parser.add_argument('-s', '--silent', action='store_true',
+            help='if set, progress bar is not printed')
+
+    args = parser.parse_args()
+    root, silent = args.root, args.silent
+
+    check_storage_format(root)
+
+    # Create a FileStorage object to use the same db settings as usual
+    file_storage = FileStorage(root)
+    db = file_storage.db
+
+    links_widgets = [
+            ' [', progress_bar.Timer(format='Time: %(elapsed)s'), '] ',
+            ' Checking links '.ljust(_ACTION_LENGTH),
+            ' ', progress_bar.Counter(), ' ',
+            progress_bar.BouncingBar()
+    ]
+
+    processed_links = 0
+    broken_links = 0
+    blob_links = {}
+
+    with progress_bar.conditional(show=not silent,
+                                  widgets=links_widgets) as bar:
+        for cur_dir, _, files in os.walk(file_storage.links_dir):
+            for file_name in files:
+                link_path = os.path.join(cur_dir, file_name)
+
+                if not os.path.islink(link_path):
+                    os.unlink(link_path)
+                    broken_links += 1
+                else:
+                    blob_path = os.path.join(
+                            os.path.dirname(link_path), os.readlink(link_path))
+                    if (os.path.islink(blob_path)
+                            or not os.path.exists(blob_path)
+                            or 'blobs/' not in blob_path):
+                        os.unlink(link_path)
+                        broken_links += 1
+                    else:
+                        digest = os.path.basename(blob_path)
+                        blob_links[digest] = blob_links.get(digest, 0) + 1
+
+                processed_links += 1
+                bar.update(processed_links)
+                import time; time.sleep(0.2)
+
+    for digest, link_count in six.iteritems(blob_links):
+        db.put(digest.encode('utf8'), str(link_count).encode('utf8'))
+
+    blobs_widgets = [
+            ' [', progress_bar.Timer(format='Time: %(elapsed)s'), '] ',
+            ' Checking blobs '.ljust(_ACTION_LENGTH),
+            ' ', progress_bar.Counter(), ' ',
+            progress_bar.BouncingBar()
+    ]
+
+    processed_blobs = 0
+    broken_blobs = 0
+
+    with progress_bar.conditional(show=not silent,
+                                  widgets=blobs_widgets) as bar:
+        for cur_dir, _, files in os.walk(file_storage.blobs_dir):
+            for blob_name in files:
+                if blob_name not in blob_links:
+                    os.unlink(os.path.join(cur_dir, blob_name))
+                    broken_blobs += 1
+
+                processed_blobs += 1
+                bar.update(processed_blobs)
+
+    if not silent:
+        print('Competed, {} broken links and {} broken blobs found.'.format(
+            broken_links, broken_blobs))
+
+
+def check_storage_format(root_dir):
```

Comment:
That's slightly pedantic, but 'check' in the function name suggests that the function will return whether or not some condition is met. But instead this function terminates the program, so maybe 'ensure' would better convey what it does?

(Feel free to ignore this comment, because it's not too important)

Reply:
I'd say it's a good point, since naming is one of the two greatest unsolved SE problems :)